Feature/setup

jmaunon opened the pull request:
# Setup configuration

## Description

It is is recommended to have a setup.py for proper version of the development.
I
This PR is related with: https://github.com/OpenMined/PyGrid/pull/544

## Type of change

Please mark options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
The existing tests should run in the same way.
The setup.py should generate a package.

## Checklist:

- [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
- [ ] New Unit tests added
- [x] Unit tests pass locally with my changes

## Progress:
I have followed the same guidelines from PySyft repository
- [x] File setup.py with description
- [x] Update of:
  - [x] tests. Now the tests uses grid package `import grid.app`
  - [x] Github actions. New line running `python setup.py install`
  - [x] Makefile. Update of make venv and make test in order to catch all requirements and run the test using coverage
- [x] Update documentation about how to install and run

Benardi commented:
It's paramount that the changes this PR brings do not break the script `GridNode/grid_node.py`.

jmaunon commented:
> It's paramount that the changes this PR brings do not break the script `GridNode/grid_node.py`.
> 

Hi. I will delete 'grid_node.py' (I forgot). The functionality of this file has been moved to `__main__.py`

Benardi commented:
![image](https://user-images.githubusercontent.com/9937551/82267951-2c671380-9944-11ea-9c6e-957205d79d37.png)

the built package does not expose any of its functions

jmaunon commented:
Hi @Benardi. I have look your comment but I do not fully understand if that is needed (or not)

I mean, at this moment the users could run the app with just `python -m grid_node <params>` (after installation). I have explained that in Readme file

Benardi commented:
> The installed package (through setup.py) doesn't provide its expected functions such as create_app
> * This type of problem is typycally solved via imports in the **init**.py

As I mentioned in my first comment the package functions need to be exposed in the **__init__.py** file, otherwise you can't import them once you have installed the package.

Benardi commented:
> > The installed package (through setup.py) doesn't provide its expected functions such as create_app
> > 
> > * This type of problem is typycally solved via imports in the **init**.py
> 
> As I mentioned in my first comment the package functions need to be exposed in the ****init**.py** file, otherwise you can't import them once you have installed the package.

Here's an [example](https://github.com/Benardi/GridNetwork/blob/pkg_setup/gridnetwork/__init__.py), notice that usually the full implementation isn't put in the __init__.py, mostly the imports.

jmaunon commented:
Thanks @Benardi. I will try to follow your recommended example. I would like all Grid family in the same standard way.

One question, in your example I did not see in `.toml` or `setup.py` any reference to `run.py` script, so if I package and distribute, the user would not that script, isn't it?

Benardi commented:
> > It's paramount that the changes this PR brings do not break the script `GridNode/grid_node.py`.
> 
> Hi. I will delete 'grid_node.py' (I forgot). The functionality of this file has been moved to `__main__.py`

@jmaunon This is not an issue as long as the functions 

> Thanks @Benardi. I will try to follow your recommended example. I would like all Grid family in the same standard way.
> 
> One question, in your example I did not see in `.toml` or `setup.py` any reference to `run.py` script, so if I package and distribute, the user would not have that script, isn't it?

Yes, it wouldn't be provided with the package

jmaunon commented:
Updated ;)

jmaunon commented:
> * [PEP guidelines](https://www.python.org/dev/peps/pep-0008/#package-and-module-names) discourage the use of underscores in package names
>   +   I would suggest using `gridnode` as the package name
> * In the module `__main__.py`some of the arguments that are set as optional should be required.
>   + If port is not provided the application breaks with a TypeError.
>   +  Should the node be started with id = None ? Unless id is a required argument this can happen.
> * In the setup.py could you set the following (to make it standard across projects) :
>   + author="OpenMined"
>   + maintainer="Benardi Nunes",
>   + maintainer_email="[benardinunes@gmail.com](mailto:benardinunes@gmail.com)"
> 
> Screenshots:
> 
> ![image](https://user-images.githubusercontent.com/9937551/82743112-d37cee00-9d3c-11ea-8db9-031c3db05f02.png)

Totally agree! We are close!

jmaunon commented:
I have noticed a side effect of this structure, I have to either specify the path of the templates or modify the folder structure.. curious...I will modify asap

jmaunon commented:
I have updated the folder st

> I have noticed a side effect of folder structure, I have to either specify the path of the templates or modify the folder structure...If not, they are not rendered I will modify asap

Fixed. I have included a `MANIFEST.in`

jmaunon commented:
Hi @IonesioJunior and @ @bobsonlin .
In this PR I have already updated to syft 0.2.6, fix ObjectStore name (see #11 ) as well and Private Tensors Allowed Users as List (see #16) however I get the following error in tests

```
==================================================================================== short test summary info ====================================================================================
FAILED test/test_api_sockets.py::test_mul_shared_tensors[x0-y0] - TypeError: can not serialize 'NodeClient' object
===================================================================== 1 failed, 34 passed, 3 skipped, 2 warnings in 23.38s ======================================================================
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:11283623780 -> james:35051590109]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:97428074686 -> james:90130656093]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:51282143813 -> james:16195955151]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:74188707713 -> james:51153022407]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:76036093559 -> bob:393112675]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:87843623547 -> alice:72988548513]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:81252377044 -> bob:89006995035]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down
Exception ignored in: <bound method ObjectPointer.__del__ of [PointerTensor | me:35741898079 -> alice:19595792190]>
Traceback (most recent call last):
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/generic/pointers/object_pointer.py", line 346, in __del__
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/workers/base.py", line 271, in send_msg
  File "/home/bda/Documents/jmaunon/GridNode/venv/lib/python3.6/site-packages/syft/serde/serde.py", line 41, in serialize
ImportError: sys.meta_path is None, Python is likely shutting down

```

jmaunon commented:
I have been trying to fix working with syft-proto versions as @karlhigley suggested in slack without sucess. 
Actually this error is not related with this PR, so we have two options:

1. Downgrade to syft 0.2.5, continue with this PR and open an issue for this error
2. Wait until is fixed

What do you think?

IonesioJunior commented:
I personally would choose to merge this PR and wait for the next syft-proto version. What do you think @Benardi?

But we still have some merge conflicts at `README.md`,` app/main/persistence/object_storage.py` and `requirements.txt`

Could you check it @Benardi, @jmaunon!

(Also, I'd like to congrats for the awesome job! Well done folks!) 🚀

jmaunon commented:
> I personally would choose to merge this PR and wait for the next syft-proto version. What do you think @Benardi?
> 
> But we still have some merge conflicts at `README.md`,` app/main/persistence/object_storage.py` and `requirements.txt`
> 
> Could you check it @Benardi, @jmaunon!
> 
> (Also, I'd like to congrats for the awesome job! Well done folks!)

I have already fixed the conflicts. According to your suggestion i have mark the test as skipped for syft.version.__version__==0.2.6 (I did not like to see test=failed in master, xd)

FYI. Syft-proto 0.4.7 has been released today but the bug is not fixed. What I do not understand is why the error  only appears in matmul and not in other operations...
